source/common/thread_local/thread_local_impl.cc

```diff
@@ -93,24 +112,24 @@ void InstanceImpl::setThreadLocal(uint32_t index, ThreadLocalObjectSharedPtr obj
 
 void InstanceImpl::shutdownGlobalThreading() {
   ASSERT(std::this_thread::get_id() == main_thread_id_);
+  ASSERT(!shutdown_);
   shutdown_ = true;
 }
 
 void InstanceImpl::shutdownThread() {
+  ASSERT(shutdown_);
+
   // Destruction of slots is done in *reverse* order. This is so that filters and higher layer
   // things that are built on top of the cluster manager, stats, etc. will be destroyed before
   // more base layer things. It's possible this might need to become more complicated later but
-  // it's OK for now.
+  // it's OK for now. Note that this is always safe to do because:
```

Comment:
The reverse ordering (or forward ordering) will not fix the pathological case that you write out above. If we get into that situation we will need to figure something out more complicated. In fact, the code that exists today will work OK with forward deletion. The reason I did reverse ordering is to deal with the potentially like case that leaf objects depend in some way on "persistent" objects (particularly the cluster manager) that are created very early on w/ a known slot number and never destroyed until the end. For example, if we chose to create persistent per-thread gRPC clients we would potentially run into shutdown issues if that thing got destroyed after the cluster manager.

Now, to your point, deletion should really be in reverse allocation order. This could be implemented relatively easily by keeping a parallel list of slot #s. But I would prefer to skip this now.

Anyhow if ^ makes sense LMK and I can add more to the comment.